Refactor for query results and failures, new protobuf protocol, fixing simulation/reachability

seblund opened the pull request:
Large PR, sorry. Almost all changes depend on each other which resulted in changes all over the code base. I will do a walkthrough on request :)

Changes:
- Update protobuf to support:
  - Simulation of all system types (Quotient works now)
  - Cause of failure for all query types with as much information as possible
- Add nice result enum types for every query and cause of failure
- Introduce intermediate types between protobuf structs and TransitionSystem related structs (State, LocationTuple, Federation, etc.) with conversions between them (`proto <=> intermediate <=> ts` ). 
- Basically rewrite simulation as the only complicated part is serialization/deserialization of states
- Change reachability to use the same `Decision` structs as in simulation

Most of the changes were to be able to serialize and deserialize `States` (Hence also `LocationTuple` and `Federation`). There are tests to ensure that the `proto <=> intermediate <=> ts` relation holds for thousands of different `States` from various systems.

t-lohse commented on src/DataReader/proto_reader.rs:
```diff
@@ -72,202 +81,175 @@ pub fn proto_state_to_state(state: ProtoState, system: &TransitionSystemPtr) ->
 }
 
 fn proto_location_tuple_to_location_tuple(
-    location_tuple: &ProtoLocationTuple,
+    location_tuple: LocationTree,
```
Since, it is no longer a `LocationTuple`, I think it would be reasonable to rename `location_tuple` to `location_tree`
```suggestion
fn proto_location_tree_to_location_tree(
    location_tree: LocationTree,
```

seblund replied:
I have renamed everything to LocationTree now. Also variables and fields so we don't reference tuples anymore 🙂

t-lohse commented on src/ProtobufServer/ecdar_requests/send_query.rs:
```diff
@@ -63,28 +63,29 @@ impl ConcreteEcdarBackend {
-
         let executable_query =
             match extract_system_rep::create_executable_query(&query, &mut component_container) {
                 Ok(query) => query,
-                Err(e) => {
+                Err(ExecutableQueryError::Custom(e)) => {
                     return Err(Status::invalid_argument(format!(
                         "Creation of query failed: {}",
                         e
                     )))
                 }
+                Err(ExecutableQueryError::SystemRecipeFailure(failure)) => {
+                    return Ok(QueryResponse {
+                        query_id: query_request.query_id,
+                        info: vec![], // TODO: Should be logs
+                        result: Some(failure.into_proto_result()),
+                    });
+                }
             };
         let result = executable_query.execute();
 
         let reply = QueryResponse {
             query_id: query_request.query_id,
             info: vec![], // TODO: Should be logs
-            result: convert_ecdar_result(&result),
+            result: Some(result.into_proto_result()),
         };
 
         Ok(reply)
```
I think this should be reduced, and avoid using the `return` keyword
```suggestion
let out =
            match extract_system_rep::create_executable_query(&query, &mut component_container) {
                Ok(query) => {
                    let result = query.execute();
                    Ok(QueryResponse {
                        query_id: query_request.query_id,
                        info: vec![], // TODO: Should be logs
                        result: Some(result.into_proto_result()),
                    })
                }
                Err(ExecutableQueryError::Custom(e)) => Err(Status::invalid_argument(format!(
                    "Creation of query failed: {}",
                    e
                ))),
                Err(ExecutableQueryError::SystemRecipeFailure(failure)) => {
                    Ok(QueryResponse {
                        query_id: query_request.query_id,
                        info: vec![], // TODO: Should be logs
                        result: Some(failure.into_proto_result()),
                    })
                }
            };
        out
```

t-lohse commented on src/ProtobufServer/ecdar_requests/send_query.rs:
```diff
@@ -142,327 +143,58 @@ fn create_components(components: Vec<Component>) -> HashMap<String, Component> {
 }
 
-fn value_in_location(maybe_location: &Option<LocationID>) -> String {
-    match maybe_location {
-        Some(location_id) => location_id.to_string(),
-        None => "".to_string(),
+impl IntoProtoResult for ConsistencyFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Consistency(self.into())
     }
 }
 
-fn value_in_component(maybe_location: Option<&LocationID>) -> Option<SpecificComponent> {
-    maybe_location.map(|location_id| SpecificComponent {
-        component_name: location_id.get_component_id().unwrap(),
-        component_index: 0,
-    })
+impl IntoProtoResult for RefinementFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Refinement(self.into())
+    }
 }
 
-fn value_in_action(maybe_action: &Option<String>) -> String {
-    match maybe_action {
-        Some(action) => action.to_string(),
-        None => "".to_string(),
+impl IntoProtoResult for PathFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Reachability(self.into())
     }
 }
```
I think that there is a lot of duplicate(ish) code here. This could be removed by using a macro instead:
```suggestion
macro_rules! impl_intoProtoResult {
    ($T:ident, $variant:ident) => {
        impl IntoProtoResult for $T {
            fn into_proto_result(self) -> ProtobufResult {
                crate::ProtobufServer::services::query_response::Result::$variant(self.into())
            }
        }
    };
}
impl_intoProtoResult!(SystemRecipeFailure, Model);
impl_intoProtoResult!(DeterminismFailure, Determinism);
impl_intoProtoResult!(ConsistencyFailure, Consistency);
impl_intoProtoResult!(RefinementFailure, Refinement);
impl_intoProtoResult!(PathFailure, Reachability);

```
While this is not a very pretty solution, I still think it is more concise. I'll let you decide :smile:

seblund replied:
I can see how the macro is more concise but i do think the macro is much less readable. I sort of don't think the code is "ugly"/long enough to warrant a macro with multiple arguments.

We could consider moving the trait implementations to a separate file?

Sort of unrelated as the duplication remains; I have now switched to implementing `From`/`Into` instead of defining the new `IntoProtoResult` trait which has the exact same functionality as `impl Into<ProtoResult>`.

t-lohse replied:
I sort of like the idea of having them in a separate file, though I feel like it is hard to justify it, so I think we should just leave it here.

t-lohse commented on src/System/extract_state.rs:
```diff
@@ -27,11 +27,12 @@ pub fn get_state(
-
-        Ok(State::create(
+        let mut state = State::create(
             build_location_tuple(&mut locations.iter(), machine, system)?,
```
Here we construct a `LocationTuple`:
1) Should we rename it to `LocationTree` as in the protobuf and,
2) Should the name of the function be `build_`..., which could imply the builder pattern or am I thinking about it to much?

This may be an issue throughout Reveaal, so maybe we should jsut make a seperate issue for it and let it go for now?

seblund replied:
I think separate issue yeah, but I agree. There are like 30+ `build_*` functions and none of them are builder patterns 😅We also use `create_*` often. I don't know what prefix (if any) we should use in general, but let's discuss in the issue.

seblund replied:
I have renamed everything to LocationTree now. Also variables so we don't reference tuples anymore 🙂

t-lohse commented on src/System/local_consistency.rs:
```diff
@@ -1,128 +1,42 @@
 pub fn is_least_consistent(system: &dyn TransitionSystem) -> ConsistencyResult {
     if system.get_initial_location().is_none() {
-        return ConsistencyResult::Failure(ConsistencyFailure::NoInitialLocation);
+        return ConsistencyFailure::no_initial_state(system);
         //TODO: figure out whether we want empty TS to be consistent
     }
 
     let mut passed = vec![];
     let state = system.get_initial_state();
     if state.is_none() {
         warn!("Empty initial state");
-        return ConsistencyResult::Failure(ConsistencyFailure::EmptyInitialState);
+        return ConsistencyFailure::no_initial_state(system);
     }
     let mut state = state.unwrap();
     state.extrapolate_max_bounds(system);
     consistency_least_helper(state, &mut passed, system)
```
I think this can be simplified:
```suggestion
    if system.get_initial_location().is_none() {
        ConsistencyFailure::no_initial_state(system)
        //TODO: figure out whether we want empty TS to be consistent
    } else if let Some(mut state) = system.get_initial_state() {
        let mut passed = vec![];
        state.extrapolate_max_bounds(system);
        consistency_least_helper(state, &mut passed, system)
    } else {
        ConsistencyFailure::no_initial_state(system)
    }
```

t-lohse commented on src/System/query_failures.rs:
```diff
@@ -0,0 +1,645 @@
+use super::specifics::{SpecificPath, SpecificState};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Composition {
```
I think it's a bit confusing having the enum called `Composition`, like the operation, and one of its fields. However, I don't have any good alternatives, maybe `Layout`, `Pattern`, `Configuration`? I'll let you decide what to do

t-lohse commented on src/System/query_failures.rs:
```diff
@@ -0,0 +1,645 @@
+        }
+    }
+
+    impl From<DeterminismFailure> for ConsistencyFailure {
+        fn from(failure: DeterminismFailure) -> Self {
+            ConsistencyFailure::NotDeterministic(failure)
+        }
+    }
+
+    impl From<SystemRecipeFailure> for QueryResult {
+        fn from(res: SystemRecipeFailure) -> Self {
+            QueryResult::RecipeFailure(res)
+        }
+    }
+
+    impl From<PathResult> for QueryResult {
+        fn from(res: PathResult) -> Self {
+            QueryResult::Reachability(res)
+        }
+    }
+
+    impl From<ConsistencyResult> for QueryResult {
+        fn from(res: ConsistencyResult) -> Self {
+            QueryResult::Consistency(res)
+        }
+    }
+
+    impl From<DeterminismResult> for QueryResult {
+        fn from(res: DeterminismResult) -> Self {
+            QueryResult::Determinism(res)
+        }
+    }
+
+    impl From<RefinementResult> for QueryResult {
+        fn from(res: RefinementResult) -> Self {
+            QueryResult::Refinement(res)
+        }
+    }
```
A lot of duplicate(ish) code here. If we implement the macro specified previously, maybe do the same here?

t-lohse commented on src/System/reachability.rs:
```diff
@@ -1,41 +1,40 @@
             return true;
         }
     }
-    // If the start state is a universal or inconsistent location and the end state isn't
-    // Since universal and inconsistent locations cannot reach other locations
-    if matches!(
-        start_state.decorated_locations.loc_type,
-        LocationType::Universal | LocationType::Inconsistent
-    ) && !start_state
-        .decorated_locations
-        .compare_partial_locations(&end_state.decorated_locations)
-    {
-        return true;
-    }
 
     false
 }
```
Can be simplified:
```suggestion
    // If any of the zones are empty
    if start_state.zone_ref().is_empty() || end_state.zone_ref().is_empty() {
        true
    }
    // If the end location has invariants and these do not have an intersection (overlap) with the zone of the end state of the query
    else if let Some(invariants) = end_state.get_location().get_invariants() {
        !end_state.zone_ref().has_intersection(invariants)
    } else {
        false
    }
```
Can even be simplified even further, but I don't think it would add anything
```rust
    // If any of the zones are empty
    start_state.zone_ref().is_empty() || end_state.zone_ref().is_empty() ||
            // If the end location has invariants and these do not have an intersection (overlap) with the zone of the end state of the query
            if let Some(invariants) = end_state.get_location().get_invariants() {
            !end_state.zone_ref().has_intersection(invariants)
        } else { false }
```

seblund replied:
I personally think the returns here are fine. The semantics are even different between your suggestion and the previous. Consider we want to add a new condition in the future:
```rust
    if start_state.zone_ref().is_empty() || end_state.zone_ref().is_empty() {
        true
    }
    else if let Some(invariants) = end_state.get_location().get_invariants() {
        !end_state.zone_ref().has_intersection(invariants)
    } else if new_cond() {
        true
    } else {
        false
    } 
```
Then `new_cond()` is never evaluated if the `end_state` has invariants, even if it has an intersection. You could of course put it before the invariant check, but are we sure that we will remember that we must do that in the future?

t-lohse replied:
That's a good point, lets just leave it as is

t-lohse commented on src/System/refine.rs:
```diff
@@ -563,114 +417,37 @@ fn prepare_init_state(
+            .map_err(|e| e.to_precondition(sys1.as_ref(), sys2.as_ref()));
+    }
+
+    Ok(())
```
Maybe chain these ifs to avoid the multiple `return`s?

t-lohse commented on src/System/specifics.rs:
```diff
@@ -0,0 +1,390 @@
+        .collect()
+}
+
+pub fn specific_clock_comp_map2(
```
Maybe rename this to show that it is a map from multiple systems? `specific_clock_comp_map_composite`?

t-lohse reviewed (changes requested):
I have a few comments, not much.

I noticed a few things that I think should be fixed/implemented/changed, but may be out of scope for this PR, such as the `LocationTuple`. Unless I have misunderstood it, I think we should rename it to `LocationTree`.
This is something I have noticed throughou Reveaal, some naming could be better. For example, if I'm not misunderstanding, a `Componenent` is the representation of the automaton? So why not just call it an `Automaton`? This way it would translate better to/from j-Ecdar. I think we should let it "slip" for this PR, and make a refactor issue.

Beyond that, I think it would be beneficial to have some documentation. Again, this is an issue throughout Reveaal, so maybe just make an issue? Personally, I think it's necessary to have documentation for public `struct`s, `enum`s, and functions, as well as "large", private helper functions, but I am not sure how much. Do you have any input?

Overall, very nice PR! :muscle:

seblund commented:
I have resolved the conversations that I have fixed with these commits. 
I have also added documentation to all structs/enums/funcs in `specifics.rs` and `query_failures.rs`. 

I have now added issues for documentation in general and for consistent naming of `build_*` methods :)

t-lohse reviewed (approved):
LGTM, very nice :smiley:
